src/wui/interactive_player.cc

```diff
@@ -774,7 +774,9 @@ bool InteractivePlayer::handle_key(bool const down, SDL_Keysym const code) {
 			set_display_flag(dfShowWorkareaOverlap, !get_display_flag(dfShowWorkareaOverlap));
 			return true;
 		}
-		if (matches_shortcut(KeyboardShortcut::kInGameScrollToHQ, code)) {
+		if (matches_shortcut(KeyboardShortcut::kInGameScrollToHQ, code) ||
+		    (get_config_bool("numpad_diagonalscrolling", false) &&
+		     code.scancode == SDL_SCANCODE_KP_5)) {
```

Comment:
Actually, I think going by scancode here is the wrong approach. I tried the following:
1. Enable diagonal scrolling
2. Assign the physical key Keypad5 (which is mapped to <kbd>=</kbd> using xkb magic) to the Open Message Menu action
3. Press <kbd>KP5</kbd>/<kbd>=</kbd> in the game → Message menu opens, keypress is ignored.
4. Reset the Message shortcut
5. Now <kbd>KP5</kbd>/<kbd>=</kbd> acts as a go-to-HQ action.

Since KP5 is a unique key on the keyboard, unlike the other KP keys which act as duplicates of other keys, it probably should not get their special treatment IMHO.

Reply:
I think the problem is that on your remapped keyboard SDL_Keysym.sym is probably '=', so it doesn't get rewritten by normalize_numpad, which looks for SDL_Keysym.sym. So to get the same effect even if the numpad is remapped on the system level, normalize_numpad should look for the scancodes as well when diagonal scrolling is requested, and change sym to the KP ones (or SDLK_UNKNOWN, as scancodes are used by the diagonal scrolling handler anyway), so 3. above would also go to HQ. This scancode based rewriting may make some sense, because the whole diagonal scrolling thing is done because of the physical layout of the keys.

Though if the code here checked for the KP5 keysym.sym, at least you'd never get the go to HQ action, ie. 5. would do nothing.
But if you remap the other KP numbers too on the system level, and assign actions that are handled before the mapview.cc handler gets its chance, then scrolling in the affected directions doesn't work either. This is what the KP 1/3/7/9 check in interactive_gamebase.cc was preventing for the default PgUp/PgDn mappings. Except that KP7 (home) is used by default here in interactive_player.cc, which never had such a test, and so stole SDLK_KP_HOME, and prevented NW scrolling. But since this bug was never reported, it looks like diagonal scrolling isn't used much anyway, so it might as well be dropped completely... ;)

Reply:
> So to get the same effect even if the numpad is remapped on the system level

But is that desired? Let's assume that I preferred to have the numpad on the left of the keyboard. So I'd remap QWE/ASD/YXC to the numpad 789/456/123 and vice versa. So I'd expect any program to treat the physical key S exactly like KP5 and the physical KP5 exactly like S. If we only check keysyms that will work fine; but if we check scancodes then it will completely ignore all my keyboard remapping. (To maximize the confusion, KP1 would have a scancode of Z and Y a scancode of KP1 because german keyboard layout).
In the case that no key with the keysym KP5 exists, like on my keyboard where it generates SDLK_EQUAL, Widelands should not pretend that this key has a function which I explicitly removed. I *could* (probably) have kept the KP5 keysym and only changed its interpretation to send the text '=' when pressed, but I chose to replace the entire key. (except when numlock is on, in which case it still generates KP5).

> But since this bug was never reported, it looks like diagonal scrolling isn't used much anyway, so it might as well be dropped completely... ;)

IIRC diagonal scrolling used to be the default, and the option was introduced because some laptop keyboards do not have PageUp/PageDown/Home keys other than the keypad ones. I'd be fine with dropping it, but we should first ask on the forum if anyone would object to that.

Reply:
> Let's assume that I preferred to have the numpad on the left of the keyboard. So I'd remap QWE/ASD/YXC to the numpad 789/456/123 and vice versa. So I'd expect any program to treat the physical key S exactly like KP5 and the physical KP5 exactly like S. If we only check keysyms that will work fine; but if we check scancodes then it will completely ignore all my keyboard remapping. (To maximize the confusion, KP1 would have a scancode of Z and Y a scancode of KP1 because german keyboard layout).

Well, in `mapview.cc` `scroll_map()` (called first from `handle_key()`), `get_key_state()` -> `SDL_GetKeyBoardState()` is used, which works with scancodes. This is required to be able to scroll diagonally by holding down 2 arrow keys, because the repeating key down events only report the last pressed key, even if some earlier key wasn't released yet. So whatever behaviour we would like to get, it needs to get down to the scancodes in `mapview.cc`.

Hmmm, looking at `SDL_keyboard.h`, it looks like there is `SDL_GetScanCodeFromKey()` to reverse look up the mapping... should I try to use that in `mapview.cc`/`scroll_map()`? (then even layout switching would work hopefully)

We could even make it to only use scancodes when numpad diagonal scrolling is off to allow diagonal scrolling with multiple keys, but turn it off and just use keysyms for the numpad when diagonal scrolling is on, so multi-key diagonal wouldn't work. Maybe multi-key could still be allowed for the main arrow keys...

> In the case that no key with the keysym KP5 exists, like on my keyboard where it generates SDLK_EQUAL, Widelands should not pretend that this key has a function which I explicitly removed. I _could_ (probably) have kept the KP5 keysym and only changed its interpretation to send the text '=' when pressed, but I chose to replace the entire key. (except when numlock is on, in which case it still generates KP5).

It seems that your key is reported as `=`, so you'd never get a KP_5 with NumLock off (no go to HQ action), but you're fine with it. Is that correct? While most users, who don't remap the numpad, would get KP_5, which is OK to be mapped to go to HQ when KP_Home (KP_7) is required for diagonal scroll. Would you accept this solution together with the above mapview change?

> > But since this bug was never reported, it looks like diagonal scrolling isn't used much anyway, so it might as well be dropped completely... ;)
> 
> IIRC diagonal scrolling used to be the default, and the option was introduced because some laptop keyboards do not have PageUp/PageDown/Home keys other than the keypad ones. I'd be fine with dropping it, but we should first ask on the forum if anyone would object to that.

Well, I didn't really mean to remove it, but the current implementation is somewhat erratic, so it's hard to figure out what I should preserve from it (while making it more predictable).

Reply:
> Hmmm, looking at SDL_keyboard.h, it looks like there is SDL_GetScanCodeFromKey() to reverse look up the mapping... should I try to use that in mapview.cc/scroll_map()? (then even layout switching would work hopefully)
>
> We could even make it to only use scancodes when numpad diagonal scrolling is off to allow diagonal scrolling with multiple keys, but turn it off and just use keysyms for the numpad when diagonal scrolling is on, so multi-key diagonal wouldn't work. Maybe multi-key could still be allowed for the main arrow keys...

Changing this dynamically would look like a bug. But reverse-lookuping keysyms to obtain their scancodes is fine, as the actual logic is then still tied to the keysym. Big +1 for that :)

> It seems that your key is reported as =, so you'd never get a KP_5 with NumLock off (no go to HQ action), but you're fine with it. Is that correct? While most users, who don't remap the numpad, would get KP_5, which is OK to be mapped to go to HQ when KP_Home (KP_7) is required for diagonal scroll. Would you accept this solution together with the above mapview change?

Correct :+1: